samza-api/src/main/java/org/apache/samza/operators/KV.java

```diff
@@ -45,4 +49,30 @@ public K getKey() {
   public V getValue() {
     return value;
   }
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    }
+    if (!(other instanceof KV)) {
+      return false;
+    }
+    KV<?, ?> otherKv = (KV<?, ?>) other;
+    return Objects.deepEquals(this.key, otherKv.key)
```

Comment:
Does this class assume key and value to be non null? If so, can we add a preconditions check in the constructor to ensure key and value are non null?

Reply:
Objects.deepEquals takes care of null checks.